### PR TITLE
Short circuit _checkUp if tap callbacks were triggered by resolve.

### PR DESCRIPTION
**Pull request by amirh**
In some scenarios this was leading to double invocation of the tap callbacks.
See detailed description in #12470.

### CODE REVIEW

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 1]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (abarth) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: Files containing tests need to end in `_test.dart` to be picked up automatically by the test runner.

**Comment by abarth** [comment]
LGTM

File: packages/flutter/lib/src/gestures/tap.dart [hunk 2]
```diff
@@ -143,6 +145,13 @@ class TapGestureRecognizer extends PrimaryPointerGestureRecognizer {
   void _checkUp() {
     if (_wonArenaForPrimaryPointer && _finalPosition != null) {
       resolve(GestureDisposition.accepted);
+      if (!_wonArenaForPrimaryPointer) {
+        // It is possible that resolve has just recursively called _checkUp (see #12470).
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/tap.dart]: assert _filePosition is also null here

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: 2017

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import 'package:flutter/widgets.dart';
+import 'package:flutter/gestures.dart';
+
+void main() {
+  testWidgets('onTap detection with canceled pointer and a drag listener', (WidgetTester tester) async {
+    int detector1TapCount = 0;
+    int detector2TapCount = 0;
+
+    final Widget widget = new GestureDetector(
+        child: new Column(
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: indent by 2

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+                  behavior: HitTestBehavior.opaque,
+                  child: const SizedBox(width: 200.0, height: 200.0),
+              ),
+              new GestureDetector(
+                onTap: () => detector2TapCount++,
+                behavior: HitTestBehavior.opaque,
+                child: const SizedBox(width: 200.0, height: 200.0)
+              )
+            ],
+          ));
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: the last parenthesis should be on its own line. the rule is if there's a newline after an open paren, there should be a newline before the close paren.

Also the indentation should be such that the close paren lines up with the first non-space character of the line that finished with the open paren.

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  testWidgets('onTap detection with canceled pointer and a drag listener', (WidgetTester tester) async {
+    int detector1TapCount = 0;
+    int detector2TapCount = 0;
+
+    final Widget widget = new GestureDetector(
+        child: new Column(
+            mainAxisAlignment: MainAxisAlignment.start,
+            children: <Widget>[
+              new GestureDetector(
+                  onTap: () => detector1TapCount++,
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: various nits:

`=>` returns the value of the expression. However, onTap is defined to return void. Technically Dart is ok with this but from a stylistic perspective we prefer to make it clear by not returning anything.

The `++` postincrement operator is generally something we avoid unless it's really what you want, because conceptually it requires storing a value temporarily. Personally I prefer `+= 1` but the preincrement operator is used by others on the team if you prefer.

In conclusion, you want:

`onTap: () { detector1TapCount += 1; },`

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+                  behavior: HitTestBehavior.opaque,
+                  child: const SizedBox(width: 200.0, height: 200.0),
+              ),
+              new GestureDetector(
+                onTap: () => detector2TapCount++,
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: ditto

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                behavior: HitTestBehavior.opaque,
+                child: const SizedBox(width: 200.0, height: 200.0)
+              )
+            ],
+          ));
+
+    await tester.pumpWidget(widget);
+
+    // The following pointer event sequence was causing the issue described
+    // in issue #12470 by triggering 2 tap events on the second detector.
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: leave an actual URL to make it easier for people to look it up

File: packages/flutter/test/widgets/gesture_disambiguation.dart [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    final TestGesture gesture1 = await tester.startGesture(const Offset(400.0, 10.0), pointer: 1);
```
Comment (Hixie) [inline packages/flutter/test/widgets/gesture_disambiguation.dart]: you shouldn't need to explicitly give the pointer IDs btw. I think they're autodetermined nowadays.

**Comment by Hixie** [comment]
Did the toString fixes not make it?

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 1 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/lib/src/gestures/tap.dart (hunk 2 above)
Comment (amirh) [reply packages/flutter/lib/src/gestures/tap.dart]: done, though not sure why is this needed here?

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 3 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 4 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 5 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 6 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 7 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 8 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

File: packages/flutter/test/widgets/gesture_disambiguation.dart (hunk 9 above)
Comment (amirh) [reply packages/flutter/test/widgets/gesture_disambiguation.dart]: done

**Comment by amirh** [comment]
@Hixie added wonArena and finalPosition to the TapGestureRecognizer's toString() (I lost the original additions we made, let me know if I missed something)

File: packages/flutter/lib/src/gestures/tap.dart [hunk 10]
```diff
@@ -159,4 +171,11 @@ class TapGestureRecognizer extends PrimaryPointerGestureRecognizer {
 
   @override
   String get debugDescription => 'tap';
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder description) {
+    super.debugFillProperties(description);
+    description.add(new FlagProperty('wonArena', value: _wonArenaForPrimaryPointer, ifTrue: 'wonArena'));
+    description.add(new DiagnosticsProperty<Offset>('finalPosition', _finalPosition));
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/tap.dart]: might as well include "sent tap down" as well while we're at it.

File: packages/flutter/lib/src/gestures/tap.dart [hunk 11]
```diff
@@ -159,4 +171,11 @@ class TapGestureRecognizer extends PrimaryPointerGestureRecognizer {
 
   @override
   String get debugDescription => 'tap';
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder description) {
+    super.debugFillProperties(description);
+    description.add(new FlagProperty('wonArena', value: _wonArenaForPrimaryPointer, ifTrue: 'wonArena'));
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/tap.dart]: probably best to say `wonArenaForPrimaryPointer` as the name, but the "ifTrue" can just be "won arena" and you can set `showName` to false (otherwise it'll say something like "wonArena: wonArena" which is confusing)

Comment (Hixie) [reply packages/flutter/lib/src/gestures/tap.dart]: oh, never mind, i see apparently FlagProperty defaults the show name thing to false.

File: packages/flutter/lib/src/gestures/tap.dart (hunk 10 above)
Comment (Hixie) [inline packages/flutter/lib/src/gestures/tap.dart]: finalPosition should have "defaultValue: null"

**Comment by amirh** [comment]
PTAL

File: packages/flutter/lib/src/gestures/tap.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/gestures/tap.dart]: Sorry the "never mind" was about the showName thing. Still best to say `wonArenaForPrimaryPointer` as the name. :-)

File: packages/flutter/lib/src/gestures/tap.dart [hunk 12]
```diff
@@ -159,4 +171,12 @@ class TapGestureRecognizer extends PrimaryPointerGestureRecognizer {
...
+    description.add(new DiagnosticsProperty<Offset>('finalPosition', _finalPosition, defaultValue: null));
+    description.add(new FlagProperty('sentTapDown', value: _wonArenaForPrimaryPointer, ifTrue: 'sentTapDown'));
```
Comment (Hixie) [inline packages/flutter/lib/src/gestures/tap.dart]: the value here is wrong... i guess that means you need to add a test to verify the debugFillProperties/toString output. :-)

File: packages/flutter/lib/src/gestures/tap.dart (hunk 11 above)
Comment (Hixie) [reply packages/flutter/lib/src/gestures/tap.dart]: for the ifTrue, let's just have `won arena` rather than camelCase

File: packages/flutter/lib/src/gestures/tap.dart (hunk 12 above)
Comment (Hixie) [reply packages/flutter/lib/src/gestures/tap.dart]: similarly here, let's have ifTrue be `sent tap down`.

**Comment by amirh** [comment]
PTAL

File: packages/flutter/test/gestures/debug_test.dart [hunk 13]
```diff
@@ -145,4 +145,14 @@ void main() {
...
   });
+
+  test('TapGestureRecognizer _sentTapDown toString', () {
+    final TapGestureRecognizer tap = new TapGestureRecognizer();
+    expect('TapGestureRecognizer#fa6ed(state: ready)', equalsIgnoringHashCodes(tap.toString()));
+    var event = const PointerDownEvent(pointer: 1, position: const Offset(10.0, 10.0));
+    tap.addPointer(event);
+    tap.didExceedDeadline();
+    expect(tap.toString(),
+      equalsIgnoringHashCodes('TapGestureRecognizer#fa6ed(state: possible, sent tap down)'));
```
Comment (Hixie) [inline packages/flutter/test/gestures/debug_test.dart]: make the hashcode in the string be #00000

File: packages/flutter/test/gestures/debug_test.dart [hunk 14]
```diff
@@ -145,4 +145,14 @@ void main() {
     debugPrintRecognizerCallbacksTrace = false;
     debugPrint = oldCallback;
   });
+
+  test('TapGestureRecognizer _sentTapDown toString', () {
+    final TapGestureRecognizer tap = new TapGestureRecognizer();
+    expect('TapGestureRecognizer#fa6ed(state: ready)', equalsIgnoringHashCodes(tap.toString()));
```
Comment (Hixie) [inline packages/flutter/test/gestures/debug_test.dart]: reverse the order of arguments, and make the hard-coded hashcode 00000

File: packages/flutter/test/gestures/debug_test.dart [hunk 15]
```diff
@@ -145,4 +145,14 @@ void main() {
...
+    var event = const PointerDownEvent(pointer: 1, position: const Offset(10.0, 10.0));
```
Comment (Hixie) [inline packages/flutter/test/gestures/debug_test.dart]: avoid var

File: packages/flutter/test/gestures/debug_test.dart (hunk 13 above)
Comment (Hixie) [reply packages/flutter/test/gestures/debug_test.dart]: (you can probably just make this one line, you don't save much by wrapping here)

**Comment by Hixie** [comment]
LGTM
